martin-mbtiles/src/tile_copier.rs

```diff
@@ -0,0 +1,202 @@
+        let storage_type = self.src_mbtiles.detect_type(&mut conn).await?;
+
+        let opt = SqliteConnectOptions::new()
+            .create_if_missing(true)
+            .filename(&self.dst_filepath);
+        let mut conn = SqliteConnection::connect_with(&opt).await?;
+
+        if query("SELECT 1 FROM sqlite_schema LIMIT 1")
+            .fetch_optional(&mut conn)
+            .await?
+            .is_some()
+        {
+            return Err(MbtError::NonEmptyTargetFile(self.dst_filepath.clone()));
+        }
+
+        query("PRAGMA page_size = 512").execute(&mut conn).await?;
+        query("VACUUM").execute(&mut conn).await?;
+
+        query("ATTACH DATABASE ? AS sourceDb")
+            .bind(self.src_mbtiles.filepath())
+            .execute(&mut conn)
+            .await?;
+
+        let schema = query("SELECT sql FROM sourceDb.sqlite_schema WHERE tbl_name IN ('metadata', 'tiles', 'map', 'images')")
+            .fetch_all(&mut conn)
+            .await?;
+
+        for row in &schema {
+            let row: String = row.get(0);
+            query(row.as_str()).execute(&mut conn).await?;
+        }
+
+        query("INSERT INTO metadata SELECT * FROM sourceDb.metadata")
+            .execute(&mut conn)
+            .await?;
+
+        match storage_type {
+            Type::TileTables => self.copy_standard_compliant_tiles(&mut conn).await,
+            Type::DeDuplicated => self.copy_deduplicated_tiles(&mut conn).await,
+        }
+    }
+
+    async fn copy_standard_compliant_tiles(&self, conn: &mut SqliteConnection) -> MbtResult<()> {
+        // TODO: Handle options
+        //  - bbox
+        //  - verbose
+        //  - zoom
+
+        self.run_query_with_options(
+            conn,
+            String::from("INSERT INTO tiles SELECT * FROM sourceDb.tiles"),
+        )
+        .await?;
+
+        Ok(())
+    }
+
+    async fn copy_deduplicated_tiles(&self, conn: &mut SqliteConnection) -> MbtResult<()> {
+        query("INSERT INTO map SELECT * FROM sourceDb.map")
+            .execute(&mut *conn)
+            .await?;
+
+        self.run_query_with_options(
+            conn,
+            String::from(
+                "INSERT INTO images
+                SELECT images.tile_data, images.tile_id
+                FROM sourceDb.images
+                  JOIN sourceDb.map
+                  ON images.tile_id = map.tile_id",
+            ),
+        )
+        .await?;
+
+        Ok(())
+    }
+
+    async fn run_query_with_options(
+        &self,
+        conn: &mut SqliteConnection,
+        mut sql: String,
+    ) -> MbtResult<()> {
+        let mut params: Vec<String> = vec![];
+
+        if !&self.options.zooms.is_empty() {
+            sql.push_str(
+                format!(
+                    " WHERE zoom_level IN ({})",
+                    vec!["?"; self.options.zooms.len()].join(",")
+                )
+                .as_str(),
+            );
+            for zoom_level in &self.options.zooms {
```

Comment:
I have seen this common oneliner for this for loop:

```rust
params.extend(self.options.zooms.iter().map(|z| z.to_string()));
```

Technically you can also do it this way, but i find it a bit less readable in some cases:

```rust
params.extend(self.options.zooms.iter().map(u8::to_string));
```